src/Bicep.Decompiler/TemplateConverter.cs

```diff
@@ -978,6 +994,35 @@ private ParameterDeclarationSyntax ParseParam(JProperty value)
                 modifier);
         }
 
+        private FunctionDeclarationSyntax ParseFunctionDeclaration(string @namespace, string name, JObject body)
+        {
+            var parameters = TemplateHelpers.AssertRequiredProperty(body, "parameters");
+            var output = TemplateHelpers.AssertRequiredProperty(body, "output");
+
+            var parametersArray = TemplateHelpers.AssertArray(parameters);
+            var outputObj = TemplateHelpers.AssertObject(output);
+
+            var parameterData = new List<(string Name, SyntaxBase Type)>();
+            foreach (var parameter in parametersArray)
+            {
+                var parameterObj = TemplateHelpers.AssertObject(parameter);
+                var parameterName = TemplateHelpers.AssertRequiredProperty(parameterObj, "name");
+                var parameterType = TemplateHelpers.AssertRequiredProperty(parameterObj, "type");
+
+                parameterData.Add((
+                    TemplateHelpers.AssertString(parameterName),
+                    TryParseType(parameterType) ?? throw new ConversionFailedException($"Unable to locate 'type' for parameter '{parameterName}'", parameterObj)));
+            }
+
+            var outputType = TryParseType(TemplateHelpers.AssertRequiredProperty(outputObj, "type")) ?? throw new ConversionFailedException($"Unable to locate 'type' for function '{name}'", body);            
+            var outputBody = PerformScopedAction(
+                () => ParseJToken(TemplateHelpers.AssertRequiredProperty(outputObj, "value")),
+                [],
+                parameterData.Select(x => x.Name));
+
+            return SyntaxFactory.CreateFunctionDeclaration($"{@namespace}_{name}", parameterData, outputType, outputBody);
```

Comment:
Should the function name be `name` instead of `$"{@namespace}_{name}"` if the namespace is `__bicep`? Given the current logic, if a template contains a function declaration like `func foo() string => 'foo'`, if you compile and then decompile it, the function will be named `__bicep_foo`; if you compile and then decompile that, then the function will be named `__bicep___bicep_foo`, etc.